cmis/src/main/java/nl/nn/adapterframework/extensions/cmis/CmisHttpInvoker.java

```diff
@@ -55,6 +55,18 @@ protected CmisHttpSender createSender() {
 		return new CmisHttpSender() {};
 	}
 
+	@Override
+	public void close() {
+		if (sender != null) {
+			try {
+				sender.close();
```

Comment:
Ik probeer te begrijpen wat het probleem is en hoe dit is onstaan.

Please, correct me if I'm wrong here.

1. Er wordt een custom CmisHttpInvoker gebruikt, een framework class
2. Deze wordt niet door de standaard SPI class gesloten
3. Door het gebruik van runtime sessies onstaat er een thread leak vanwege het niet sluiten van de custom CmisHttpInvoker

De oplossing is:

1. Custom SPI class dat wel de close aanroept
2. De sender correct sluiten in de custom CmisHttpInvoker

Zou er in dit geval niet een issue/pr report naar het Apache CMIS project gemaakt kunnen worden om bij het sluiten van de standaard SPI class te controleren of de CmisHttpInvoker een auto-closable instance is en deze dan ook te sluiten?

Misschien zit ik er compleet naast hoor, maar dan hoor ik graag waar de denkfout zit :)

Reply:
Een PR voor Apache CMIS? Zou kunnen. Maar dan zou je misschien de hele BindingSession willen nalopen op AutoCloseables.

En deze patch heb je morgen. Een patch for OpenCMIS? 🤔 
De laatste release op mvnrepository is van 2017.

De analyse is verder juist. Wanneer runtime-requests worden gebruikt, wordt voor ieder request een nieuwe CMIS session gemaakt. Die maakt een Apache HttpClient (via de invoker, via de CmisHttpSender). De Apache HttpClient heeft in het Frank!Framework een standaard configuratie met eigen ConnectionManager en IdleConnectionEvictor thread en CMIS `session.clear()` doet niets, en `bindings.close()` eigenlijk ook niet.
Dus de HttpClient wordt uiteindelijk wel opgeruimd door de garbage collector, maar de ConnectionManager en het Connection evictor thread niet.

Het duurde even voordat ik het doorgrond had.

Reply:
Het was zeker een lastige, wij kwamen zelf ook niet verder dan veel threads maar waarom... ons vermoeden lag meer in de JMS hoek. We zijn erg blij met je onderzoek, bedankt voor de toelichting verder :)

Verder, het Apache CMIS project lijkt vrij dood te zijn, jammer.

Kan je misschien ook toelichten hoe dit probleem is ontstaan, of waarom dit pas aan het licht kwam bij het upgraden naar deze versie? Als het veel werk is om dit uit te zoeken, laat het dan maar achterwege, puur stukje interesse

Reply:
Dat was ik inderdaad vergeten toe te voegen.

In F!F 7.7 werden op de HttpClientBuilder nog niet de properties gezet waardoor de IdleConnectionEvictor thread wordt gemaakt.

Dit lijkt verder niet voor problemen te zorgen omdat HttpClients verder wel normaal worden vastgehouden en gesloten.

Doordat dit probleem alleen optrad met CMIS en alleen met specifieke configuratie settings, is het verder in geen enkele test opgevallen.

Ik heb overigens daardoor ook nog niet een goede test-case die ik kan toevoegen aan het F!F.

Reply:
Nogmaals bedankt, dat is een duidelijk verhaal :)

Ik zal dit morgen ook binnen het team delen.

In het verleden hebben we ook andere issues ervaren die pas optraden bij een redelijke hoeveelheid load. Naar mijn mening is het een goed idee om voor de integratie testen ook een load/performance test variant op te stellen waarin de verschillende implementaties getest worden.

Elke build lijkt mij dit een beetje een overkill gezien de duratie van dit soort testen vrij lang kan zijn. Maar eens per week de running release te nemen of bij het uitbrengen van een release dit type testen te draaien lijkt mij een goede toevoeging.

Wij beschikken al over enkele JMeter gebasseerde testen en zijn ook bereid om te helpen bij het opstellen van deze testen.

Op deze manier zou je bij een bepaalde spec van de host machine een bepaalde performance kunnen garanderen en daarnaast vang je ook mogelijke leaks, hoe denken jullie hierover?

Reply:
We denken daar af en toe wel eens over maar tot nu toe hebben we hiervoor nog niet de tijd gevonden, ivm prioriteiten op andere features, bug fixes, en capaciteit binnen het team!

Het probleem wat we nu hadden zou ook niet zo snel gevonden zijn met load tests omdat het afhankelijk is van vrij specifieke configuratie opties, en op dit moment hebben we volgens mij nog niet eens een Larva test scenario voor CMIS dus daar moet eerst aan gewerkt worden.

De unit tests hebben niet genoeg integratie tussen verschillende componenten om het daarin goed te kunnen testen.

Reply:
Begrijpelijk, ik zal ook binnen het team voorstellen om deze verantwoordelijkheid voorlopig op ons te nemen betreft de CMIS sender.

In principe hebben we al load testen, we zullen deze moeten configureren om ook z'n 25K berichten aan te tikken per service en zullen deze aftrappen op het moment dat we upgrades doen.

Mochten jullie in de toekomst zover zijn om ook dit soort testen zelf te hosten, voel je vrij om contact op te nemen, hopelijk hebben we tegen die tijd zelf het eea staan en kunnen we dit overdragen.

Reply:
Bedankt!

Reply:
Fijn dat jullie dat willen doen @Laurens-makel . Wij hebben volgens mij nog wel wat huiswerk/achterstand in überhaupt loadtests op gewoon de basis-set configurations in F!F. We meten dit niet structureel, zover ik weet. Ik denk dat we eerst daarmee bezig moeten en daarna specifiek iets voor CMIS. Jullie hebben die systemen al, dus fijn als je dat per grote release wil checken.